### PR TITLE
Improvements for wxSTC autocompletion and call tips

### PR DESCRIPTION
**Pull request by NPDevelopmentCode**
As I mentioned on the dev list, I realize this is a large pull request; but I think the net result is a better behaving popup class for autocompletion lists and call tips.  This should also address all outstanding issues related to autocompletion.

- Commits 1 and 2 are additions to the stc sample.  The first adds a demonstration of autocompletion for C preprocessor directives.  The second demonstrates handling clicks in call tips by using 3 related call tips and letting the user cycle between the tips by clicking up or down buttons.

newcalltip

- Commit 3 adds the wxSTCPopupWindow class with customizations for windows and GTK+ to make sure it has the behavior it needs.  The class is based on the popup class currently used so on any other port the behavior should be the same as it is now.

- The next four commits (4 through 7) implement the wxSTCPopupWindow for wxCocoa.  The first of these adds an objective C file with several functions needed for the popup class and uses them in the C++ code.  Since I had to add a new source file and header, the next commit updates the wxWidgets build system files to use the new files.  Commit 6 is the result of running upmake and bakefile.  The final commit of this set adds the new file to the xcode project.  This is obviously unfotunate, but I could not find any other way to get a popup to not take focus on macos any other way.

- Commits 8 and 9 use the new popup window to show autocompletion lists and call tips respectively.

- Commit 10 introduces a custom listbox class based on wxVListBox.  The custom listbox is built specifically to never take focus but always look like it has focus.  This commit also adds a helper class to keep track of the listbox’s images and colors.

- Commits 11 through 15 are enhancements for the popup classes. 

  * Commits 11 and 12 let a popup window move when its parent stc window moves and hide when its application window is minimized.

  * Commit 13 is windows specific and is intended to reduce flicker when switching between different call tips.  I had to switch the drawing behavior to cache the calltip background in a bitmap and draw the bitmap in the erase background event and add a very quick fade in effect for call tip windows.  I would have thought either of those would have been sufficient, but I had to do both to get rid of the flicker.

  * Commit 14 adds a method to wxSTC to let the colors be specified for an autocompletion window. Commit 15 adds a method to make the autocompletion list look like it’s being shown with a list control instead of a listbox.  I added this because that’s how autocompletion lists are currently shown, and this lets a user maintain the current style if they want to.  Also, I think on windows it just looks better. 

autocomplistctrl

- Commit 16 is just regenerating the stc files after changes where made stc.cpp.in and stc.h.in.

- Commits 17 through 19 add a gui test to verify the behavior of the stc popups is correct.  Commit 17 adds the test file and modifies the build system files.  Commit 18 is the result of running bakefile.  The final commit updates the remaining build files to link with the stc library.

I hope the breakdown of commits is helpful and I won't be offended if it takes a while to evaluate this pull request.

### CODE REVIEW

**Comment by NewPagodi** [comment]
Oops.  I guess I need to add the wxScintilla lib for the gui test with the builds using static libs.  I need to ask for help since I don't know conditionally add that lib for static builds.

**Comment by vadz** [comment]
Sorry, what's the actual problem? I see some compilation errors due to missing headers in Travis, but I'm not sure what's wrong with the libraries?

**Comment by NewPagodi** [comment]
4 of the build failures were due to some mistakes I made with the CMake files.  I temporarily removed the commits with the gui test so I could focus on those.  
Now that I the CMake issues fixed, I'll try to get the test working tomorrow.

**Comment by NewPagodi** [comment]
I got the test working.  I had to make 2 changes.  First I surrounded the test with a check for `WXUSINGDLL` as discussed above.  Unfortunately, that means the test will only be run with the builds using shared libraries, but that's better than nothing.

Second, to get the test working with the 2 Travis cases where `--disable-stc` is set, I changed the definition of `WXLIB_STC` in common.bkl from

`<if cond="MONOLITHIC=='0'">$(mk.evalExpr(wxwin.mkLibName('stc')))</if>`

to

`<if cond="MONOLITHIC=='0' and USE_STC=='1'">$(mk.evalExpr(wxwin.mkLibName('stc')))</if>`

and then in test.bkl, I listed the link library as `<sys-lib>$(WXLIB_STC)</sys-lib>` as is currently done for the webview library.  That resulted in a change to autoconf_inc.m4 and the contributing docs say to run autoconf if that happens.  I did that, but I'm only following the directions and don't really understand what that does.

In short, I think this is right, but the second to last commit should probably be looked at closely.

**Comment by vadz** [comment]
Thanks again for all your work on this!

I'm afraid I don't really understand why does the test only in the DLL builds, but I agree that it's not a critical problem. As for `WXLIB_STC` change, it definitely seems correct to me as we're already doing the same thing for the other libraries.

I'll try to review this, but it will be a cursory review at most and I don't expect to find any real problems, so this could be merged soon. However I'd like to keep it open for at least a while (a week?) and hope that people using wxSTC in their applications could test this PR and leave their comments if they find any problems.

File: src/stc/PlatWX.cpp [hunk 1]
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
...
+#else
 
-    void OnFocus(wxFocusEvent& event) {
-        GetParent()->SetFocus();
-        event.Skip();
-    }
+    wxSTCPopupBase::wxSTCPopupBase(wxWindow* parent):wxFrame()
+    {
+        // Make sure the frame is initially hidden. However, GTK+ will hide the
+        // frame initially and doesn't like trying to hide it before it's
```
Comment (vadz) [inline src/stc/PlatWX.cpp]: This is pretty bad actually, hiding any window before creating it is definitely supposed to work.

OTOH frames are also supposed to be created hidden under all platforms, so this call shouldn't be _necessary_ anywhere. It should't do any harm however neither...

File: samples/stc/edit.cpp [hunk 2]
```diff
@@ -169,6 +186,14 @@ Edit::Edit (wxWindow *parent, wxWindowID id,
     // annotations
     AnnotationSetVisible(wxSTC_ANNOTATION_BOXED);
 
+    // autocompletion
+    wxBitmap bmp(hashtag_xpm);
+    RegisterImage(0, bmp);
+
+    //call tips
```
Comment (vadz) [inline samples/stc/edit.cpp]: Very nitpicky:
```suggestion
    // call tips
```

File: samples/stc/edit.cpp [hunk 3]
```diff
@@ -481,11 +504,50 @@ void Edit::OnCharAdded (wxStyledTextEvent &event) {
...
+    else if (chr == '#') {
+        wxString s = "define?0 elif?0 else?0 endif?0 error?0 if?0 ifdef?0 "
+                     "ifndef?0 include?0 line?0 pragma?0 undef?0";
+        AutoCompShow(0,s);
+    }
+}
+
+void Edit::OnCallTipClick(wxStyledTextEvent &event)
+{
+    if ( event.GetPosition() == 1 ) {
```
Comment (vadz) [inline samples/stc/edit.cpp]:
```suggestion
    if ( event.GetPosition() == 1 )
    {
```
(also in a few more places below)

File: src/stc/PlatWX.cpp [hunk 4]
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
...
+                }
+            }
+            else
+                return wxFrame::Show(false);
+        }
 
-};
+        // Do not activate in response to mouse clicks on this window.
+        bool wxSTCPopupBase::MSWHandleMessage(WXLRESULT *res, WXUINT msg,
+                                              WXWPARAM wParam, WXLPARAM lParam)
```
Comment (vadz) [inline src/stc/PlatWX.cpp]:
```suggestion
                                              WXWPARAM wParam, WXLPARAM lParam) wxOVERRIDE
```

File: src/stc/PlatWX.cpp [hunk 5]
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
...
+        #endif
     }
 
-    wxListView* GetLB() { return lv; }
+    #ifdef __WXMSW__
 
-private:
-    wxDECLARE_EVENT_TABLE();
+        // Use ShowWithoutActivating instead of show.
+        bool wxSTCPopupBase::Show(bool show)
```
Comment (vadz) [inline src/stc/PlatWX.cpp]:
```suggestion
        bool wxSTCPopupBase::Show(bool show) wxOVERRIDE
```

File: src/stc/PlatWX.cpp [hunk 6]
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
...
+    #endif
+
+        if ( !m_highlightTextColourIsSet )
+            m_highlightTextColour =
+                wxSystemSettings::GetColour(wxSYS_COLOUR_LISTBOXHIGHLIGHTTEXT);
     }
+}
 
-    wxListView* GetLB() { return lv; }
+void SetColourHelper(bool& isSet, wxColour& itemCol, const wxColour& newColour)
```
Comment (vadz) [inline src/stc/PlatWX.cpp]:
```suggestion
static void SetColourHelper(bool& isSet, wxColour& itemCol, const wxColour& newColour)
```
(or put it an unnamed namespace)

File: src/stc/PlatWX.cpp [hunk 7]
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
...
+
+    int      m_desiredVisibleRows;
+    ImgList  m_imgList;
+
+    wxColour m_borderColour;
+    wxColour m_bgColour;
+    wxColour m_textColour;
+    wxColour m_highlightBgColour;
+    wxColour m_highlightTextColour;
+    bool     m_bgColourIsSet;
```
Comment (vadz) [inline src/stc/PlatWX.cpp]: Why do we need these `xxxIsSet` variables? It seems strange to store the colour validity outside of it when we could just use `xxx.IsOk()` instead.

File: src/stc/PlatWX.cpp [hunk 8]
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
...
+            // style, the colours used seem to be based on the parent's
+            // background colour. So set the popup's background.
+            parent->SetOwnBackgroundColour(m_visualData->GetBgColour());
+        #endif
+    }
 }
 
+wxSTCListBox::~wxSTCListBox()
+{
+    m_labels.clear();
```
Comment (vadz) [inline src/stc/PlatWX.cpp]: This is not really needed, why manually do something done automatically anyhow?

File: src/stc/PlatWX.cpp [hunk 9]
```diff
@@ -2418,37 +2742,52 @@ PRectangle ListBoxImpl::GetDesiredRect() {
...
+{
+    m_doubleClickAction = action;
+    m_doubleClickActionData = data;
 }
 
-void ListBoxImpl::SetList(const char* list, char separator, char typesep) {
-    GETLB(wid)->Freeze();
+void wxSTCListBox::SetList(const char* list, char separator, char typesep)
+{
+    Freeze();
```
Comment (vadz) [inline src/stc/PlatWX.cpp]: Using `wxWindowUpdateLocker` is safer.

File: src/stc/PlatWXcocoa.mm [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+// A simple view used for popup windows.
+
+@interface wxSTCPopupBaseView : NSView
+{
+@private
+    NSTrackingArea * m_trackingArea;
+    wxWindow*        m_wxWin;
+}
+
+- (id)initWithwxWin:(wxWindow*) wxwin;
```
Comment (vadz) [inline src/stc/PlatWXcocoa.mm]:
```suggestion
- (id)initWithwxWin:(wxWindow*) wxWin;
```
just for consistency

**Review by vadz — APPROVED** [review]
The promised week turned into a month, but, finally, I did review the code and, as expected, I don't have any real comments about it, just a few nitpicks here and there.

The amount of platform-specific code needed to achieve what you did is truly dispiriting, but I won't even try to understand what could we do to encapsulate the differences in wx because I know I won't have time to do it anyhow (and I suspect neither will you), so we'll just have to live with this.

I can certainly merge it as is, but if you could please address the comments below, it would be great. If you won't be able to do it in the near future, please let me know.

Thanks again for all your work on this!

File: samples/stc/edit.cpp (hunk 3 above)
Comment (NewPagodi) [reply samples/stc/edit.cpp]: This was to match the code style in the surrounding methods. Should I change it to match the current code guidelines?

File: src/stc/PlatWX.cpp (hunk 7 above)
Comment (NewPagodi) [reply src/stc/PlatWX.cpp]: The IsSet variables are tracking when the colors are set by the user.  An, admittedly exotic, example where they might be needed is in responding to system color changes.  When the `ComputeColours` function gets called, we need to recompute all the colors that have not been specifically set by the user.  At that point it won't be sufficient to check if the color is OK because all the colors will be OK due to a previous to the `ComputeColours` function.

I know there were other cases where those variables were needed, but that's the only one I remember right now.  I agree they're ugly and I tried to get rid of them, but I couldn't get everything to work without them.

File: samples/stc/edit.cpp (hunk 3 above)
Comment (NewPagodi) [reply samples/stc/edit.cpp]: I'll have to make a new commit to address the other 3 comments.  Should I change these at that time as well or should I leave them to match the surrounding methods?

Comment (vadz) [reply samples/stc/edit.cpp]: I haven't realized this was to match the existing style. I don't really know what to do in such cases, perhaps the best would be to commit this as is and then reformat it all at once in a later commit? In any case, please feel free to ignore this suggestion.

File: src/stc/PlatWX.cpp (hunk 7 above)
Comment (vadz) [reply src/stc/PlatWX.cpp]: Sorry, I forgot to answer this one: I thought that perhaps we could avoid `xxxIsSet` by just checking if the colour is valid in `GetXXX()` and returning its default value (used in `ComputeColours()`) if it isn't.

If this doesn't work for some reason I'm missing, it's not a big deal to keep the `xxxIsSet`, of course, but I think they should be renamed to `xxxIsSetByUser` or, maybe, invert its meaning and call it `useDefaultXXX` because currently "is set" is rather confusing: it can be false even if the colour is, actually, set to a valid value.

Comment (NewPagodi) [reply src/stc/PlatWX.cpp]: I've changed the names to m_useDefaultXXXColour.  That probably is a better name for the purpose those variables serve.

File: src/stc/PlatWX.cpp (hunk 1 above)
Comment (NewPagodi) [reply src/stc/PlatWX.cpp]: I just removed the call to Hide everywhere since I now know it's done by default everywhere.

File: src/stc/PlatWX.cpp (hunk 8 above)
Comment (NewPagodi) [reply src/stc/PlatWX.cpp]: This has been removed.

File: src/stc/PlatWX.cpp (hunk 9 above)
Comment (NewPagodi) [reply src/stc/PlatWX.cpp]: This has been changed,

**Comment by vadz** [comment]
Merged, thanks again for all your work on this!

**Comment by vadz** [comment]
@NewPagodi Unfortunately I'm seeing test failures with the tests added here, see e.g. https://travis-ci.org/wxWidgets/wxWidgets/jobs/559896693 but I also can reproduce this locally without Xvfb: both `AutoComp` and `Calltip` tests fail sporadically for me (or, rather, pass sporadically, as they fail more often than not, especially with GTK2, but even with GTK3 I still see the failures often enough too). I've tried debugging this, but without much success...

For now I think I'll just disable these tests under GTK, but it would be nice to fix them, especially if this means that the popup actually doesn't work correctly.

**Comment by NewPagodi** [comment]
Well that's frustrating.  GTK+ was the only port where things seemed to work out of the box. I think the location of the mouse clicks must miscalculated and partly random for some reason so that they sometimes landing outside the application window.  I don't think the popup is really taking focus, because if it was, the test would fail 100% of the time.
